pkg/controller/middleware/sessions.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/logging"
+	"github.com/google/exposure-notifications-verification-server/pkg/render"
+	"github.com/gorilla/sessions"
+	"go.uber.org/zap"
+)
+
+const (
+	// sessionName is the name of the session.
+	sessionName = "verification-server-session"
+)
+
+// RequireSession retrieves or creates a new session and stores it on the
+// request's context for future retrieval. It also ensures the flash data is
+// populated in the template map. Any handler that wants to utilize sessions
+// should use this middleware.
+func RequireSession(ctx context.Context, store sessions.Store, h *render.Renderer) func(http.Handler) http.Handler {
+	logger := logging.FromContext(ctx).Named("middleware.RequireSession")
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+
+			// Get or create a session from the store.
+			session, err := store.Get(r, sessionName)
+			if err != nil {
+				// We couldn't get a session (invalid cookie, can't talk to redis,
+				// whatever). According to the spec, this can return an error but can never
+				// return an empty session. We intentionally discard the error to ensure we
+				// have a session.
+				session, _ = store.New(r, sessionName)
+			}
+
+			// Save the flash in the template map.
+			m := controller.TemplateMapFromContext(ctx)
+			m["flash"] = controller.Flash(session)
+			ctx = controller.WithTemplateMap(ctx, m)
+
+			// Save the session on the context.
+			ctx = controller.WithSession(ctx, session)
+			*r = *r.WithContext(ctx)
+
+			// Ensure the session is saved at least once. This is passed to the
+			// before-first-byte writer AND called after the middleware executes to
+			// ensure the session is always sent.
+			var once sync.Once
+			saveSession := func() error {
+				var err error
+				once.Do(func() {
+					session := controller.SessionFromContext(ctx)
+					if session != nil {
+						err = session.Save(r, w)
+					}
+				})
+				return err
+			}
+
+			bfbw := &beforeFirstByteWriter{
+				w:      w,
+				before: saveSession,
+				logger: logger,
+			}
+
+			logger.Debugw("done")
```

Comment:
fine to have debug log statement, but "done" probably isn't helpful on it's own

Reply:
It won't be "done", it'll be `middleware.RequireSession: done`. Named loggers are pretty nice: https://github.com/google/exposure-notifications-verification-server/pull/171/files#diff-0ad52782f7ddd2e9fd3f45181ca4eb7fR39